lib/src/source/hosted.dart

```diff
@@ -1100,3 +1113,97 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+extension GCSChecksumValidation on http.StreamedResponse {
+  static const checksumsHeaderName = 'x-goog-hash';
+
+  bool get isFromGCS {
+    // TODO: Does this suffice as a heuristic?
+    return headers.containsKey(checksumsHeaderName);
+  }
+
+  /// Adds a checksum validation "tap" to the response stream and returns a
+  /// wrapped Stream object, which should be used to consume the incoming data.
+  ///
+  /// As chunks are received, a CRC32C checksum is updated.
+  /// Once the download is completed, the final checksum is compared with
+  /// the one present in the `x-goog-hash` response header from GCS.
+  ///
+  /// Throws [PackageIntegrityException] if anything is wrong with the checksum
+  /// or if there is a checksum mismatch.
+  Stream<List<int>> streamWithGCSChecksumValidationTap() {
+    if (gcsCrc32c == null) {
+      throw PackageIntegrityException(
+          'Package served from GCS is missing CRC32C checksum', request?.url);
+    }
+
+    final checksumComputer = Crc32c();
+
+    return stream
+        .transform(onDataTransformer(checksumComputer.update))
+        .transform(onDoneTransformer(() {
+      var computedCrc32c = checksumComputer.finalize();
+
+      log.io(
+          'Computed checksum ($computedCrc32c) for package with GCS checksum of ($gcsCrc32c).');
+
+      if (gcsCrc32c != computedCrc32c) {
+        throw PackageIntegrityException(
+            'Package served from GCS has a CRC32C checksum mismatch; Computed checksum ($computedCrc32c) != GCS checksum ($gcsCrc32c)',
+            request?.url);
+      }
+    }));
+  }
+
+  /// Parses GCS response headers and returns the package's CRC32C checksum.
+  int? get gcsCrc32c {
+    String? gcsChecksums = headers[checksumsHeaderName];
+    if (gcsChecksums == null) return null;
+
+    // In most cases, GCS provides both MD5 and CRC32C checksums in its response
+    // headers. It uses the header name "x-goog-hash" for these values. It has
+    // been documented and observed that GCS will send multiple response headers
+    // with the same "x-goog-hash" token as the key.
+    // https://cloud.google.com/storage/docs/xml-api/reference-headers#xgooghash
+
+    // Additionally, when the Dart http package encounters multiple response
+    // headers with the same key, it concatenates their values with a comma
+    // before inserting a single item with that key and concatenated value into
+    // its response "headers" Map. There is a TODO to refactor the
+    // response class to use a HttpHeaders object instead of a Map, but the
+    // person it was assigned to, nweiz@, has long since left the Dart team.
+    // https://github.com/dart-lang/http/issues/24
+    // https://github.com/dart-lang/http/blob/06649afbb5847dbb0293816ba8348766b116e419/pkgs/http/lib/src/base_response.dart#L29
```

Comment:
I like that we have some extra details, but no reason to call out usernames,
```suggestion
    // Additionally, when the Dart http client encounters multiple response
    // headers with the same key, it concatenates their values with a comma
    // before inserting a single item with that key and concatenated value into
    // its response "headers" Map. 
    // See https://github.com/dart-lang/http/issues/24
    // https://github.com/dart-lang/http/blob/06649afbb5847dbb0293816ba8348766b116e419/pkgs/http/lib/src/base_response.dart#L29
```

Reply:
Yea, I wasn't sure on what the proper etiquette was so I erred towards more information. Noted for the future, though!